File: test/windows_conpty.zig
const std = @import("std");
const Repo = @import("git_repo.zig").Repo;
const pager_path = @import("test_options").pager_path;
const runConPty = @import("windows_conpty_support.zig").runConPty;

test "Git pager opens and closes a native ConPTY without changing the repository" {
    var memory = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer memory.deinit();
    const arena = memory.allocator();
    var repo = try Repo.init(arena, std.testing.io);
    defer repo.deinit();
    try repo.write("Example.cs", "before\n");
    try repo.commit();
    try repo.write("Example.cs", "after\n");
    try repo.setPager(pager_path, false);
    const status_before = try repo.git(&.{ "status", "--porcelain=v1" });
    const config_before = try repo.read(".git/config");
    const result = try runConPty(arena, std.testing.io, &repo);
    try std.testing.expect(result.sent_quit);
    try std.testing.expect(result.exit_code == 0);
    try std.testing.expect(std.mem.indexOf(u8, result.transcript, "before") != null);
    try std.testing.expect(std.mem.indexOf(u8, result.transcript, "after") != null);
    try std.testing.expect(std.mem.indexOf(u8, result.transcript, "\x1b[?1049h") != null);
    try std.testing.expect(std.mem.indexOf(u8, result.transcript, "\x1b[?1049l") != null);
    try std.testing.expectEqualStrings(status_before, try repo.git(&.{ "status", "--porcelain=v1" }));
    try std.testing.expectEqualStrings(config_before, try repo.read(".git/config"));
    try std.testing.expectEqualStrings("after\n", try repo.read("Example.cs"));
    const executable = try std.Io.Dir.cwd().realPathFileAlloc(std.testing.io, pager_path, arena);
    const empty = try std.process.run(arena, std.testing.io, .{ .argv = &.{executable} });
    try std.testing.expectEqual(@as(u8, 0), empty.term.exited);
    try std.testing.expectEqualStrings("", empty.stdout);
}

File: test/windows_conpty.py
#!/usr/bin/env python3
"""Prove Git's pager pipe and console input work through native ConPTY."""

import os
import re
import select
import shlex
import subprocess
import sys
import tempfile
import time
from pathlib import Path

from winpty import Backend, PtyProcess


ANSI = re.compile(r"\x1b\[[0-?]*[ -/]*[@-~]")
QUERIES = {
    "\x1b[6n": "\x1b[1;1R",
    "\x1b[5n": "\x1b[0n",
    "\x1b[?u": "\x1b[?0u\x1b[?1;2c",
}


def git(repo, *args):
    return subprocess.check_output(["git", *args], cwd=repo)


def run_pager(repo):
    environment = os.environ.copy()
    environment.pop("GIT_PAGER", None)
    environment.pop("PAGER", None)
    environment["TERM"] = "xterm-256color"

    # pywinpty treats integer zero as an absent backend, so the string forces ConPTY.
    process = PtyProcess.spawn(
        ["git", "--paginate", "diff"],
        cwd=str(repo),
        env=environment,
        dimensions=(24, 80),
        backend=str(Backend.ConPTY),
    )
    transcript = ""
    responses = {query: 0 for query in QUERIES}
    sent_quit = False
    deadline = time.monotonic() + 20
    try:
        while time.monotonic() < deadline:
            readable, _, _ = select.select([process.fileobj], [], [], 0.1)
            if readable:
                try:
                    transcript += process.read(65536)
                except EOFError:
                    break
                for query, reply in QUERIES.items():
                    count = transcript.count(query)
                    while responses[query] < count:
                        process.write(reply)
                        responses[query] += 1
                visible = ANSI.sub("", transcript)
                # The screen header proves Git has fed the patch before the key is sent.
                if "Before (-)" in visible and "Example.cs" in visible and not sent_quit:
                    process.write("q")
                    sent_quit = True
            if not process.isalive() and "\x1b[?1049l" in transcript:
                break
        else:
            raise AssertionError(f"ConPTY pager timed out: {transcript[-1200:]!r}")
        assert sent_quit, transcript[-1200:]
        assert process.wait() == 0, transcript[-1200:]
        assert "before" in transcript and "after" in transcript
        assert "\x1b[?1049h" in transcript, "viewer did not enter alternate screen"
        assert "\x1b[?1049l" in transcript, "viewer did not leave alternate screen"
    finally:
        process.close(force=True)


def main():
    executable = Path(sys.argv[1]).resolve()
    assert executable.is_file(), executable
    assert Backend.ConPTY == 0
    with tempfile.TemporaryDirectory(prefix="lantana-conpty-") as directory:
        repo = Path(directory)
        git(repo, "init", "-q")
        git(repo, "config", "user.name", "Lantana Test")
        git(repo, "config", "user.email", "test@example.invalid")
        git(repo, "config", "commit.gpgsign", "false")
        (repo / "Example.cs").write_text("before\n", encoding="utf-8")
        git(repo, "add", "Example.cs")
        git(repo, "commit", "-qm", "initial")
        (repo / "Example.cs").write_text("after\n", encoding="utf-8")
        git(repo, "config", "core.pager", shlex.quote(executable.as_posix()))
        git(repo, "config", "pager.diff", "true")
        git(repo, "config", "color.ui", "false")
        patch = git(repo, "--no-pager", "diff")
        assert patch
        status_before = git(repo, "status", "--porcelain=v1")
        config_before = (repo / ".git" / "config").read_bytes()
        run_pager(repo)
        assert git(repo, "status", "--porcelain=v1") == status_before
        assert (repo / ".git" / "config").read_bytes() == config_before
        assert (repo / "Example.cs").read_text(encoding="utf-8") == "after\n"
        empty = subprocess.run([str(executable)], input=b"", capture_output=True, timeout=5)
        assert empty.returncode == 0 and empty.stdout == b""


if __name__ == "__main__":
    main()
